### PR TITLE
Add support to attach filter onto socket on Linux

### PR DESCRIPTION
**Pull request by hhyhhy**
Hi

This patch add support to attach filter onto socket on Linux.
See also: [Linux Socket Filtering aka Berkeley Packet Filter (BPF)](https://www.kernel.org/doc/html/v5.12/networking/filter.html)

### CODE REVIEW

File: src/sys/unix.rs [hunk 1]
```diff
@@ -200,6 +205,31 @@ type IovLen = usize;
 ))]
 type IovLen = c_int;
 
+/// Required for SO_ATTACH_FILTER.
+#[cfg(all(feature = "all", target_os = "linux"))]
+#[derive(Debug)]
+#[repr(C)]
+struct SockFprog {
```
Comment (Thomasdezeeuw) [inline src/sys/unix.rs:212]: Please us `libc` https://docs.rs/libc/0.2.106/libc/struct.sock_fprog.html.

File: src/sys/unix.rs [hunk 2]
```diff
@@ -200,6 +205,31 @@ type IovLen = usize;
...
+    ///  Number of filter blocks
+    pub len: c_ushort,
+    pub filter: *mut SockFilter,
+}
+
+/// Linux Filter block
+#[cfg(all(feature = "all", target_os = "linux"))]
+#[derive(Debug)]
+#[repr(C)]
+pub struct SockFilter {
```
Comment (Thomasdezeeuw) [inline src/sys/unix.rs:222]: Please us `libc` https://docs.rs/libc/0.2.106/libc/struct.sock_filter.html.

File: src/lib.rs [hunk 3]
```diff
@@ -131,6 +131,9 @@ compile_error!("Socket2 doesn't support the compile target");
 
 use sys::c_int;
 
+#[cfg(all(feature = "all", target_os = "linux"))]
+pub use sys::SockFilter;
```
Comment (Thomasdezeeuw) [inline src/lib.rs:135]: Don't expose this.

File: src/sys/unix.rs [hunk 4]
```diff
@@ -1773,6 +1803,41 @@ impl crate::Socket {
...
         }
     }
+
+    /// Attach Berkeley Packet Filter(BPF) on this socket.
+    ///
+    /// BPF allows a user-space program to attach a filter onto any socket
+    /// and allow or disallow certain types of data to come through the socket.
+    ///
+    /// For more information about this option, see [filter](https://www.kernel.org/doc/html/v5.12/networking/filter.html)
+    #[cfg(all(feature = "all", target_os = "linux"))]
```
Comment (Thomasdezeeuw) [inline src/sys/unix.rs:1813]: Does Android support this?

File: src/sys/unix.rs [hunk 5]
```diff
@@ -1773,6 +1803,41 @@ impl crate::Socket {
...
+    pub fn attach_filter(&self, mut filters: Vec<SockFilter>) -> io::Result<()> {
```
Comment (Thomasdezeeuw) [inline src/sys/unix.rs:1814]: We don't need ownership of the `filter`s, use `&[libc::sock_filter]` instead and we can create the `libc::sock_fprog` structure based on that.

File: src/sys/unix.rs [hunk 6]
```diff
@@ -1773,6 +1803,41 @@ impl crate::Socket {
...
+            filter: filters.as_mut_ptr(),
+        };
+        forget(filters);
+
+        unsafe {
+            setsockopt(
+                self.as_raw(),
+                libc::SOL_SOCKET,
+                libc::SO_ATTACH_FILTER,
+                prog,
```
Comment (Thomasdezeeuw) [inline src/sys/unix.rs:1829]: Pretty sure this needs to be `&prog`.

**Review by Thomasdezeeuw — CHANGES_REQUESTED** [review]

File: src/sys/unix.rs (hunk 6 above)
Comment (hhyhhy) [reply src/sys/unix.rs:1829]: The payload has been converted to a pointer at: 

https://github.com/rust-lang/socket2/blob/5c7b03f61ef9ddb21233a18f643d6b9367429c74/src/sys/unix.rs#L963

I tried to change to `&prog` and it did not run successfully as before.

File: src/sys/unix.rs (hunk 4 above)
Comment (hhyhhy) [reply src/sys/unix.rs:1813]: According to the filter [documentation](https://android.googlesource.com/kernel/msm/+/refs/tags/android-4.4w_r10/Documentation/networking/filter.txt) in android kernel, it should be supported.

I will fix it.

File: src/sys/unix.rs [hunk 7]
```diff
@@ -1773,6 +1773,37 @@ impl crate::Socket {
...
     }
+
+    /// Attach Berkeley Packet Filter(BPF) on this socket.
+    ///
+    /// BPF allows a user-space program to attach a filter onto any socket
+    /// and allow or disallow certain types of data to come through the socket.
+    ///
+    /// For more information about this option, see [filter](https://www.kernel.org/doc/html/v5.12/networking/filter.html)
+    #[cfg(all(feature = "all", any(target_os = "linux", target_os = "android")))]
+    pub fn attach_filter(&self, filters: &mut [libc::sock_filter]) -> io::Result<()> {
```
Comment (Thomasdezeeuw) [inline src/sys/unix.rs:1784]: We don't need a mutable reference I think? I assume the kernel is only reading the values.

You can use `filters.as_ptr() as *mut _`.

File: src/sys/unix.rs (hunk 6 above)
Comment (Thomasdezeeuw) [reply src/sys/unix.rs:1795]: You're right, I should really try to remember the code I wrote :)

**Review by Thomasdezeeuw — COMMENTED** [review]
Final small change.

**Comment by Thomasdezeeuw** [comment]
It seems aarch64-linux-android doesn't have `libc::SO_ATTACH_FILTER`. We should add that to the libc crate.

**Comment by hhyhhy** [comment]
I have created a [pr](https://github.com/rust-lang/libc/pull/2506) for libc.

**Comment by Thomasdezeeuw** [comment]
I think this now depends on https://github.com/rust-lang/libc/pull/2506.

**Comment by Thomasdezeeuw** [comment]
@hhyhhy if you can update to  libc v0.2.107 this can be merged 👍

**Comment by hhyhhy** [comment]
It seems that the problem on the Android system has been solved.

**Comment by Thomasdezeeuw** [comment]
Thanks @hhyhhy

**Pull request merged**